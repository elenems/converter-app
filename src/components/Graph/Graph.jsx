import React, { Component } from "react";
import { connect } from "react-redux";
import Chart from "./Chart";

class Graph extends Component {
  shouldComponentUpdate(nextProps, nextState) {
    if (
      this.props.historyRates !== nextProps.historyRates ||
      this.props.isLoading !== nextProps.isLoading
    ) {
      return true;
    }
    return false;
  }

  render() {
    const canvasStyle = this.props.isLoading
      ? { display: "none" }
      : { display: "block" };
    const paragraphStyle = this.props.isLoading
      ? {
          display: "block",
          fontSize: "24px",
          fontWeight: "bold",
          color: "#3f51b5"
        }
      : { display: "none" };
    return (
      <div style={{ textAlign: "center" }}>
        <p style={paragraphStyle}>Loading...</p>
        <div style={canvasStyle} className="chart-container">
          <Chart
            currencyFrom={this.props.currencyFrom}
            currencyTo={this.props.currencyTo}
            historyRates={this.props.historyRates}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    historyRates: state.currency.historyRates,
    currencyFrom: state.currency.currencyFrom,
    currencyTo: state.currency.currencyTo,
    historyFrom: state.currency.historyFrom,
    isLoading: state.ui.isLoading
  };
};

export default connect(mapStateToProps)(Graph);
