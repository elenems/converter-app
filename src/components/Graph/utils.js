export const getCurrencyData = (currencies, from, to) => {
  const differences = [];
  const dates = [];
  let sortedCurrencies = [];

  for (let curr in currencies) {
    sortedCurrencies.push({
      date: curr,
      currencies: currencies[curr]
    });
  }

  sortedCurrencies = sortedCurrencies.sort((a, b) => {
    return new Date(a.date) - new Date(b.date);
  });

  for (let x of sortedCurrencies) {
    if (x["currencies"][from] !== x["currencies"][to]) {
      differences.push(x["currencies"][to] - x["currencies"][from]);
    } else {
      differences.push(x["currencies"][from]);
    }
    dates.push(x["date"]);
  }
  return { differences: differences, dates: dates };
};
