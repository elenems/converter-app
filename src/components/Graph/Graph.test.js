import { shallow } from "enzyme";
import React from "react";
import { GraphTest } from "./Graph.jsx";
import Graph from "../Graph";
import configureMockStore from "redux-mock-store";
describe("Graph", () => {
  const mockStore = configureMockStore();
  const initialState = {
    currency: {
      amountFrom: 1,
      amountTo: 0,
      currencyFrom: "",
      currencyTo: "",
      rates: null,
      historyRates: [],
      historyFrom: new Date("2015-01-01"),
      historyTo: new Date()
    },
    ui: {
      error: "",
      isLoading: false
    }
  };
  it("renders", () => {
    const wrapper = shallow(<GraphTest />).debug();
    expect(wrapper).toMatchSnapshot();
  });

  it("should update component", () => {
    const nextProps = {
        isLoading:true
    }
    const wrapper = shallow(<GraphTest isLoading={false}/>);
    const mock = jest.spyOn(wrapper.instance(), 'shouldComponentUpdate');
    wrapper.instance().shouldComponentUpdate(nextProps)
    expect(mock).toHaveBeenCalled();
  });

  it("should not update component", ()=>{
    const nextProps = {
        isLoading:true
    }
    const wrapper = shallow(<GraphTest isLoading={true}/>);
    const mock = jest.spyOn(wrapper.instance(), 'shouldComponentUpdate');
    wrapper.instance().shouldComponentUpdate(nextProps)
    expect(mock).toHaveBeenCalled();
  })

  it("connects state", () => {
    const store = mockStore(initialState);
    const wrapper = shallow(<Graph store={store} />);
    expect(wrapper.find("Graph").props()).toBeTruthy();
  });
});
