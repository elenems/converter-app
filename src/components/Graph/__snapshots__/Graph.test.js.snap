// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph renders 1`] = `
"<div style={{...}}>
  <p style={{...}}>
    Loading...
  </p>
  <div style={{...}} className=\\"chart-container\\">
    <Chart currencyFrom={[undefined]} currencyTo={[undefined]} historyRates={[undefined]} />
  </div>
</div>"
`;
